### ISSUE TITLE
Is streaming source data possible?

### ISSUE THREAD

**Issue by fenom** [czheo/syntax_sugar_python#18]
Is streaming from, say, a database cursor possible? I don't want to have to load everything before data starts moving through the pipe.

**Comment by czheo** [comment]
Surely, pipe is lazily evaluated.

**Comment by fenom** [comment]
https://github.com/czheo/syntax_sugar_python/blob/c68000248de0f267062589f8af07b4d2b26fb221/syntax_sugar/_pipe.py#L67
I found that I had to change this line to use `imap` and `list`, just like in `multigreenthread`, to get my desired behavior. `map` consumes the entire database cursor before running the function, and I couldn't afford that. My function writes back to the database and discards the data, so even though the `IMapIterator` was consumed immediately anyway, it was much better than loading all the data first. And I wonder if it's possible to make parallel processing fully lazy by yielding from the `IMapIterator` and making generator syntax with parentheses, that way the parallel processing stage doesn't have to be the final stage. Square brackets can be kept as eager syntax, just like list comprehension.
